File: lc_go/lcGo/backtrack/permute.go
package main

func permute(nums []int) [][]int {
	ans := [][]int{}
	res := []int{}
	n := len(nums)
	vis := make([]bool, n)
	for i := range vis {
		vis[i] = false
	}
	var dfs func(i int)
	dfs = func(i int) {
		if i == n {
			tmp := make([]int, len(res))
			copy(tmp, res)
			ans = append(ans, tmp)
			return
		}

		for j := 0; j < n; j++ {
			if !vis[j] {
				res = append(res, nums[j])
				vis[j] = true
				dfs(i + 1)
				res = res[:len(res)-1]
				vis[j] = false
			}
		}
	}
	dfs(0)
	return ans
}

func main() {

}

File: lc_py/lcPy/backtrack/permute.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans, res, n = [], [], len(nums)
        vis = [False] * n
        def backTrack(i: int) -> None:
            if i == n:
                ans.append(res[:])  # 创建 res 的副本
                return
            for j in range(n):  # 修复：用 j 遍历，避免覆盖 i
                if not vis[j]:
                    res.append(nums[j])
                    vis[j] = True
                    backTrack(i + 1)
                    res.pop()
                    vis[j] = False
        backTrack(0)
        return ans